inc/class-tmsc-sync.php

```diff
@@ -356,43 +385,84 @@ public function do_post_processing() {
-													'value'   => $constituent_ids,
+													'value'   => $ids,
 													'compare' => 'IN',
 												),
 											),
 										) );
-										if ( ! empty( $related_constituents ) ) {
-											$meta_data[ $role_slug ] = $related_constituents;
+
+										if ( ! empty( $related_posts ) && $existing_relationships !== $related_posts ) {
+											update_post_meta( $post_id, $key, $related_posts );
+										}
+									} elseif ( 'constituent' === $relationship_map[ $key ]['type'] ) {
+										foreach ( $ids as $constituent_type_slug => $role_data ) {
+											$meta_data = array();
+											foreach( $role_data as $role_slug => $constituent_ids ) {
+												if ( ! empty( $constituent_ids ) ) {
+													$related_constituents = get_posts( array(
+														'fields' => 'ids',
+														'suppress_filters' => false,
+														'ignore_sticky_posts' => true,
+														'no_found_rows' => true,
+														'post_type' => $relationship_map[ $key ]['slug'],
+														'post_status' => 'publish',
+														'meta_query' => array(
+															array(
+																'key'     => 'tmsc_legacy_id',
+																'value'   => $constituent_ids,
+																'compare' => 'IN',
+															),
+														),
+													) );
+													if ( ! empty( $related_constituents ) ) {
+														$meta_data[ $role_slug ] = $related_constituents;
+													}
+												}
+											}
+											if ( ! empty( $meta_data ) ) {
+												update_post_meta( $post_id, $constituent_type_slug, $meta_data );
+											}
+										}
+									} elseif ( 'link' === $relationship_map[ $key ]['type'] ) {
+										$link_meta = array();
+										if ( apply_filters( 'tmsc_enable_links', false ) ) {
+											// TODO: Bookmark Integration
+										} else {
+											foreach ( $ids as $resource_data ) {
+												$link_meta[] = array(
+													'link' => $resource_data,
+												);
+											}
+											if ( ! empty( $link_meta ) ) {
+												update_post_meta( $post_id, $key, $link_meta );
+											}
 										}
 									}
 								}
-								if ( ! empty( $meta_data ) ) {
-									update_post_meta( $post_id, $constituent_type_slug, $meta_data );
-								}
-							}
-						} elseif ( 'link' === $relationship_map[ $key ]['type'] ) {
-							$link_meta = array();
-							if ( apply_filters( 'tmsc_enable_links', false ) ) {
-								// TODO: Bookmark Integration
-							} else {
-								foreach ( $ids as $resource_data ) {
-									$link_meta[] = array(
-										'link' => $resource_data,
-									);
-								}
-								if ( ! empty( $link_meta ) ) {
-									update_post_meta( $post_id, $key, $link_meta );
-								}
 							}
 						}
+						if ( apply_filters( 'tmsc_delete_post_processing_data', false ) ) {
+							delete_post_meta( $post_id, 'tmsc_post_processing' );
+						}
+					}
+
+					tmsc_stop_the_insanity();
```

Comment:
Should this function only be called in a `WP_CLI` context?

Reply:
Yes, this script may be called via a cron job and in any context we should prevent memory leaks.